File: OOP/container.cpp
#include "container.h"
#include "film.h"
#include <cstring>
#include "drama.h"
#include "documentary.h"
#include "cartoon.h"

char* ReadString(FILE*);

container::container(int length) {
    this->length = length;
    films = new film *[length];
}

container::~container() {
    for (int i = 0; i < length; ++i) {
        delete films[i];
    }

    delete[] films;
}

// Read container from file.
void container::Read(FILE *file) {
    fscanf(file, "%d", &length);

    films = new film *[length];
    for (int i = 0; i < length; ++i) {
		char* type = ReadString(file);

		if (strcmp(type, "drama") == 0) {
			films[i] = new drama();
		} else if (strcmp(type, "cartoon") == 0) {
			films[i] = new cartoon();
		} else if (strcmp(type, "documentary") == 0) {
			films[i] = new documentary();
		} else {
			exit(1);
		}

    	films[i]->Read(file);
        free(type);
    }
}

// Write container to file.
void container::Write(FILE *file) {
    fprintf(file, "%d%s", length, "\n");
    for (int i = 0; i < length; ++i) {
        films[i]->Write(file);
        fprintf(file, "%s", "\n");
    }
}

// Create random container with [size] films.
void container::Random(int size) {
    length = size;
    films = new film *[length];

    for (int i = 0; i < size; ++i) {
        int type = rand() % 3;
        if (type == 0) {
            films[i] = new drama();
        } else if (type == 1) {
            films[i] = new documentary();
        } else if (type == 2) {
            films[i] = new cartoon();
        }

        films[i]->Random();
    }

}

// Sort container.
void container::ShellSort() {
    for (int s = length / 2; s > 0; s /= 2) {
        for (int i = 0; i < length; i++) {
            for (int j = i + s; j < length; j += s) {
                if (films[i]->Calculate() > films[j]->Calculate()) {
                    film *temp = films[i];
                    films[i] = films[j];
                    films[j] = temp;
                }
            }
        }
    }
}

File: Dynamic/Container.py
from Drama import Drama
from Cartoon import Cartoon
from Documentary import Documentary
from Utilities import random_int


class Container:
    def __init__(self, length):
        self.length = length
        self.films = list()

    @staticmethod
    def read(file):
        length = int(file.readline())
        container = Container(length)

        for i in range(0, length):
            params = file.readline().split()
            if params[0] == 'drama':
                film = Drama()
            elif params[0] == 'cartoon':
                film = Cartoon()
            else:
                film = Documentary()

            film.read(params)
            container.films.append(film)

        return container

    def write(self, file):
        file.write(str(self.length) + '\n')

        for film in self.films:
            file.write(film.to_string() + '\n')

    @staticmethod
    def random(length):
        container = Container(length)

        for i in range(0, length):
            film_type = random_int(3)

            if film_type == 0:
                film = Drama()
            elif film_type == 1:
                film = Cartoon()
            else:
                film = Documentary()

            film.random()
            container.films.append(film)

        return container

    def shell_sort(self):
        s = self.length // 2

        while s > 0:
            for i in range(0, self.length):

                j = i + s
                while j < self.length:
                    if self.films[i].calculate() > self.films[j].calculate():
                        temp = self.films[i]
                        self.films[i] = self.films[j]
                        self.films[j] = temp
                    j += s

            s //= 2
